### PR TITLE
Fix several logical errors in arrayFold

### LINKED ISSUE #59809: arrayFold: Logical error: 'Cannot capture column 3 because it has incompatible type: got Int64, but LowCardinality(Int64) is expected.'
Report: https://s3.amazonaws.com/clickhouse-test-reports/0/548cc2577b41f04a93494205a54e8506ff2b230c/ast_fuzzer__debug_.html

```sql
SELECT arrayFold((acc, x) -> (acc + (x * 2)), [1, 2, 3, 4], toInt64(toLowCardinality(3)))
```

2024.02.09 15:55:03.320496 [ 358270 ] {48e9a48c-0c6b-4bbf-8a10-14cb03de7b4e} <Fatal> : Logical error: 'Cannot capture column 3 because it has incompatible type: got Int64, but LowCardinality(Int64) is expected.'.

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix several logical errors in arrayFold. Fixes support for Nullable and LowCardinality

### Documentation entry for user-facing changes

Closes https://github.com/ClickHouse/ClickHouse/issues/59809
No memory problems/invalid accesses because of them, it always reached an exception.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 91f54f44b60cd0c9ac21495f1c801a6297512ad8 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60022/91f54f44b60cd0c9ac21495f1c801a6297512ad8/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60022/91f54f44b60cd0c9ac21495f1c801a6297512ad8/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
How does default implementation for NULLs/LowCardinality introduce logical errors in this case?
Shouldn't the default implementation handle it correctly anyway?

**Comment by Algunenano** [comment]
> How does default implementation for NULLs/LowCardinality introduce logical errors in this case?
> Shouldn't the default implementation handle it correctly anyway?

Good question. The problem is that the validation that happens in getReturnTypeImpl with default implementation is without NULLs or LC (they are removed), but the lambda function is keeping them. I just realized there is a `getLambdaArgumentTypes` for each function, so I'll have a look if that's where the problem is (as getLambdaArgumentTypes should get the same types as getReturnTypeImpl`.

**Comment by Algunenano** [comment]
After looking into it there isn't a simpler solution that to ignore the default null and lc adaptors.

The problem is the following:
* In order to prepare the arguments of the arrayFold, as it has a lambda argument, `getLambdaArgumentTypes` is called with the original types (that include nullable or LC). With this we can search for the final function, knowing its input and more importantly what would be its return type.
* Once we have the lambda adapted and the columns, we have the ability to create the final function and know it's return type (for which we need to pass the input types).
* The problem comes from the fact that the default adaptors will modify the arguments of the functions so that they don't have to deal with NULLs or LowCardinality, but when we generated the lambda function we didn't know anything about this change, so either they won't match here or they won't match when executing the function. We could try changing the adaptors, but that doesn't seem possible since it requires knowing what the function will do to the arguments too (in this case transforms the arrays into columns for example).

So the simpler option is to have all the logic directly in the function. I'll just add a comment and a couple of extra tests

File: src/Functions/array/arrayFold.cpp [hunk 1]
```diff
@@ -32,6 +32,9 @@ class FunctionArrayFold : public IFunction
     size_t getNumberOfArguments() const override { return 0; }
     bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
 
+    /// Avoid the default adaptors since they modify the inputs and that makes knowing the lambda argument types
```
Comment (alexey-milovidov) [inline src/Functions/array/arrayFold.cpp:35]: Ok, but we can also do the following:
- default implementations are ineffective in the presence of lambda functions as arguments.

**Comment by clickgapai** [comment]
Hi @Algunenano @alexey-milovidov — while reviewing this PR I found the following:

- **Coverage gap:** Non-const `Array(LowCardinality(T))` path in `arrayFold::executeImpl` is untested — https://github.com/ClickHouse/ClickHouse/pull/103861

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**